lib/galaxy/model/store/__init__.py

```diff
@@ -1222,20 +1256,22 @@ def __init__(
             serialize_files_handler=self,
         )
         self.export_files = export_files
-        self.included_datasets = {}
-        self.included_collections = []
-        self.included_libraries = []
-        self.included_library_folders = []
-        self.collection_datasets = {}
-        self.collections_attrs = []
-        self.dataset_id_to_path = {}
+        self.included_datasets: Dict[model.DatasetInstance, Tuple[model.DatasetInstance, bool]] = {}
+        self.included_collections: List[Union[model.DatasetCollection, model.HistoryDatasetCollectionAssociation]] = []
+        self.included_libraries: List[model.Library] = []
+        self.included_library_folders: List[model.LibraryFolder] = []
+        self.collection_datasets: Set[int] = set()
```

Comment:
Should we do
```suggestion
        self.collection_datasets: Dict[model.DatasetInstance, Tuple[model.DatasetInstance, bool]] = {}
```
? I guess right now we only expect HDAs, but if we ever allow other things as collection datasets we might run into odd bugs ?

Reply:
Python illiterate question: do we have to redefine `__hash__()` and `__eq__()` in `model.DatasetInstance` to safely use the class as dict key? or it 'just works'?

Reply:
@mvdbeek Well right now we only use this to do one thing (a dict we place True into for some integer IDs - which wold be better represented as a set). If we want to use it differently in the future like you suggest we can just change it then right? It is only used in a few places and we have excellent test coverage.

Reply:
I don't think tests would catch a mistake like that (identical primary keys for different items is what I'm worried about), but fair enough, we'll probably remember to do that.